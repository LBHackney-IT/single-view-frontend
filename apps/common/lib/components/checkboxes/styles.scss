@import "~lbh-frontend/lbh/base";
@import "~lbh-frontend/lbh/components/lbh-checkboxes/checkboxes";
