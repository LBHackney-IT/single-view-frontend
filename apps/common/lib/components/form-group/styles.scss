@import "../../../node_modules/lbh-frontend/lbh/base";
@import "../../../node_modules/govuk-frontend/govuk/objects/form-group";
@import "../../../node_modules/lbh-frontend/lbh/objects/form-group";
@import "../../../node_modules/lbh-frontend/lbh/components/lbh-label/label";
@import "../../../node_modules/lbh-frontend/lbh/components/lbh-hint/hint";
@import "../../../node_modules/lbh-frontend/lbh/components/lbh-error-message/error-message";

.lbh-form-group {
  label,
  legend {
    sup {
      color: lbh-colour("lbh-error");
    }
  }
}

fieldset.lbh-form-group legend {
  display: contents;
  margin-bottom: 0.375rem;
}
