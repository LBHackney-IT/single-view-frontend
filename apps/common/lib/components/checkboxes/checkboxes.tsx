import React, {
  Children,
  ComponentPropsWithoutRef,
  ReactNode,
  forwardRef,
  isValidElement,
  useEffect,
  useMemo,
  useRef,
} from "react";
import mergeRefs from "react-merge-refs";

import cn from "classnames";
import { Checkboxes } from "lbh-frontend";

import "./styles.scss";

export interface CheckboxProps extends ComponentPropsWithoutRef<"input"> {
  id: string;
  hint?: string;
  children: ReactNode;
  conditionalId?: string;
  error?: string;
}

export const Checkbox = forwardRef<HTMLInputElement, CheckboxProps>(function Checkbox(
  { id, className, type = "checkbox", hint, children, conditionalId, error, ...props },
  ref,
) {
  return (
    <div className={cn("govuk-checkboxes__item", className)}>
      <input
        ref={ref}
        id={id}
        className="govuk-checkboxes__input"
        type={type}
        aria-describedby={hint ? `${id}-hint` : undefined}
        data-aria-controls={conditionalId}
        {...props}
      />
      <label className="govuk-label govuk-checkboxes__label" htmlFor={id}>
        {children}
      </label>
      {hint ? (
        <span id={`${id}-hint`} className="govuk-hint govuk-checkboxes__hint lbh-hint">
          {hint}
        </span>
      ) : null}
    </div>
  );
});

export const CheckboxConditional = forwardRef<
  HTMLDivElement,
  ComponentPropsWithoutRef<"div">
>(function CheckboxConditional(props, ref) {
  return (
    <div
      ref={ref}
      className="govuk-checkboxes__conditional govuk-checkboxes__conditional--hidden"
      {...props}
    />
  );
});

export interface CheckboxGroupProps extends ComponentPropsWithoutRef<"div"> {
  variant?: "base" | "small";
  error?: string;
}

export const CheckboxGroup = forwardRef<HTMLDivElement, CheckboxGroupProps>(
  function CheckboxGroup({ variant = "base", children, error, ...props }, ref) {
    const localRef = useRef<HTMLDivElement>();

    useEffect(() => {
      /* istanbul ignore else */
      if (localRef.current) {
        new Checkboxes(localRef.current).init();
      }
    }, []);

    const hasConditionals = useMemo(
      () =>
        Children.toArray(children).some(
          (child) => isValidElement(child) && child.type === CheckboxConditional,
        ),
      [children],
    );

    return (
      <div
        ref={mergeRefs([localRef, ref])}
        className={cn(
          "govuk-checkboxes",
          {
            "govuk-checkboxes--small": variant === "small",
            "govuk-checkboxes--conditionals": hasConditionals,
          },
          "lbh-checkboxes",
        )}
        {...props}
      >
        {children}
      </div>
    );
  },
);
