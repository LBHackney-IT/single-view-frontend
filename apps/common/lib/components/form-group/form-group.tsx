import React, {
  Children,
  ReactElement,
  cloneElement,
  forwardRef,
  isValidElement,
  useMemo,
} from "react";

import classNames from "classnames";

import { widthOverrides } from "../../utils";
import { TextArea } from "../text-area";

import type * as Polymorphic from "@radix-ui/react-polymorphic";
import "./styles.scss";

export interface FormGroupProps {
  id: string;
  label?: string;
  name?: string;
  hint?: string;
  error?: string | false;
  required?: boolean;
  children: ReactElement;
  override?: number;
}

export type FormGroupComponent = Polymorphic.ForwardRefComponent<"div", FormGroupProps>;

export const FormGroup: FormGroupComponent = forwardRef(function FormGroup(
  {
    as: FormGroupComp = "div",
    id,
    name,
    label,
    hint,
    error,
    required,
    children,
    className,
    override,
    ...props
  },
  ref,
) {
  const formGroupClasses = classNames(
    "govuk-form-group",
    {
      "govuk-form-group--error": !!error,
    },
    "lbh-form-group",
    widthOverrides(override),
    className,
  );

  const describedBy = useMemo(() => {
    const classes: string[] = [];
    if (hint) {
      classes.push(`${id}-hint`);
    }
    if (error) {
      classes.push(`${id}-error`);
    }
    return classes.join(" ");
  }, [id, hint, error]);

  const LabelComp =
    typeof FormGroupComp === "string" && (FormGroupComp as string) === "fieldset"
      ? "legend"
      : "label";

  const formGroup = (
    <FormGroupComp ref={ref} id={id} className={formGroupClasses} {...props}>
      <LabelComp className="govuk-label lbh-label" htmlFor={`${id}-field`}>
        {label}
        {required ? <sup aria-hidden="true">*</sup> : ""}
      </LabelComp>
      {!!hint && (
        <span id={`${id}-hint`} className="govuk-hint lbh-hint">
          {hint}
        </span>
      )}
      {!!error && (
        <span id={`${id}-error`} className="govuk-error-message lbh-error-message">
          <span className="govuk-visually-hidden">Error:</span> {error}
        </span>
      )}
      {!!children &&
        Children.only(
          cloneElement(children, {
            id: `${id}-field`,
            name,
            required,
            error: !!error,
            "aria-describedby": describedBy || undefined,
            ...children.props,
          }),
        )}
    </FormGroupComp>
  );

  return isValidElement(children) && children.type === TextArea ? (
    <div className="govuk-character-count">{formGroup}</div>
  ) : (
    formGroup
  );
});
