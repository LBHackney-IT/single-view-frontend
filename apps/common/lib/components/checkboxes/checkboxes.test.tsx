import React from "react";

import { render, testA11y } from "@hackney/mtfh-test-utils";

import { FormGroup } from "../form-group";
import { Checkbox, CheckboxConditional, CheckboxGroup } from "./checkboxes";

test("it passes a11y on single checkbox", async () => {
  const { container } = render(
    <FormGroup id="test" name="test" label="Checkbox Test">
      <Checkbox id="radio">Label</Checkbox>
    </FormGroup>,
  );
  expect(container).toMatchSnapshot();
  await testA11y(container);
});

test("it passes a11y on checkbox group", async () => {
  const { container } = render(
    <FormGroup id="test" label="Checkbox Test">
      <CheckboxGroup>
        <Checkbox id="checkbox-1" name="test-1">
          Label
        </Checkbox>
        <Checkbox id="checkbox-2" name="test-2">
          Label2
        </Checkbox>
      </CheckboxGroup>
    </FormGroup>,
  );
  expect(container).toMatchSnapshot();
  await testA11y(container);
});

test("it renders a checkbox", () => {
  const { container } = render(<Checkbox id="test">Label</Checkbox>);
  expect(container).toMatchSnapshot();
});

test("it renders a checkbox with a hint", () => {
  const { container } = render(
    <Checkbox id="test" hint="A short description">
      Label
    </Checkbox>,
  );
  expect(container).toMatchSnapshot();
});

test("it renders a group of checkboxes", () => {
  const { container } = render(
    <CheckboxGroup>
      <Checkbox id="test">Label</Checkbox>
      <Checkbox id="test2">Label2</Checkbox>
    </CheckboxGroup>,
  );
  expect(container).toMatchSnapshot();
});

test("it renders a group of checkboxes with a conditional", () => {
  const { container } = render(
    <CheckboxGroup>
      <Checkbox id="test" conditionalId="conditional">
        Label
      </Checkbox>
      <CheckboxConditional id="conditional">Hello</CheckboxConditional>
      <Checkbox id="test2">Label2</Checkbox>
    </CheckboxGroup>,
  );
  expect(container).toMatchSnapshot();
});
