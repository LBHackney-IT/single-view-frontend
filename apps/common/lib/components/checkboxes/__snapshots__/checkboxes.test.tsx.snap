// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it passes a11y on checkbox group 1`] = `
<div>
  <div
    class="govuk-form-group lbh-form-group"
    id="test"
  >
    <label
      class="govuk-label lbh-label"
      for="test-field"
    >
      Checkbox Test
      
    </label>
    <div
      class="govuk-checkboxes lbh-checkboxes"
      id="test-field"
    >
      <div
        class="govuk-checkboxes__item"
      >
        <input
          class="govuk-checkboxes__input"
          id="checkbox-1"
          name="test-1"
          type="checkbox"
        />
        <label
          class="govuk-label govuk-checkboxes__label"
          for="checkbox-1"
        >
          Label
        </label>
      </div>
      <div
        class="govuk-checkboxes__item"
      >
        <input
          class="govuk-checkboxes__input"
          id="checkbox-2"
          name="test-2"
          type="checkbox"
        />
        <label
          class="govuk-label govuk-checkboxes__label"
          for="checkbox-2"
        >
          Label2
        </label>
      </div>
    </div>
  </div>
</div>
`;

exports[`it passes a11y on single checkbox 1`] = `
<div>
  <div
    class="govuk-form-group lbh-form-group"
    id="test"
  >
    <label
      class="govuk-label lbh-label"
      for="test-field"
    >
      Checkbox Test
      
    </label>
    <div
      class="govuk-checkboxes__item"
    >
      <input
        class="govuk-checkboxes__input"
        id="radio"
        name="test"
        type="checkbox"
      />
      <label
        class="govuk-label govuk-checkboxes__label"
        for="radio"
      >
        Label
      </label>
    </div>
  </div>
</div>
`;

exports[`it renders a checkbox 1`] = `
<div>
  <div
    class="govuk-checkboxes__item"
  >
    <input
      class="govuk-checkboxes__input"
      id="test"
      type="checkbox"
    />
    <label
      class="govuk-label govuk-checkboxes__label"
      for="test"
    >
      Label
    </label>
  </div>
</div>
`;

exports[`it renders a checkbox with a hint 1`] = `
<div>
  <div
    class="govuk-checkboxes__item"
  >
    <input
      aria-describedby="test-hint"
      class="govuk-checkboxes__input"
      id="test"
      type="checkbox"
    />
    <label
      class="govuk-label govuk-checkboxes__label"
      for="test"
    >
      Label
    </label>
    <span
      class="govuk-hint govuk-checkboxes__hint lbh-hint"
      id="test-hint"
    >
      A short description
    </span>
  </div>
</div>
`;

exports[`it renders a group of checkboxes 1`] = `
<div>
  <div
    class="govuk-checkboxes lbh-checkboxes"
  >
    <div
      class="govuk-checkboxes__item"
    >
      <input
        class="govuk-checkboxes__input"
        id="test"
        type="checkbox"
      />
      <label
        class="govuk-label govuk-checkboxes__label"
        for="test"
      >
        Label
      </label>
    </div>
    <div
      class="govuk-checkboxes__item"
    >
      <input
        class="govuk-checkboxes__input"
        id="test2"
        type="checkbox"
      />
      <label
        class="govuk-label govuk-checkboxes__label"
        for="test2"
      >
        Label2
      </label>
    </div>
  </div>
</div>
`;

exports[`it renders a group of checkboxes with a conditional 1`] = `
<div>
  <div
    class="govuk-checkboxes govuk-checkboxes--conditionals lbh-checkboxes"
  >
    <div
      class="govuk-checkboxes__item"
    >
      <input
        aria-controls="conditional"
        aria-expanded="false"
        class="govuk-checkboxes__input"
        id="test"
        type="checkbox"
      />
      <label
        class="govuk-label govuk-checkboxes__label"
        for="test"
      >
        Label
      </label>
    </div>
    <div
      class="govuk-checkboxes__conditional govuk-checkboxes__conditional--hidden"
      id="conditional"
    >
      Hello
    </div>
    <div
      class="govuk-checkboxes__item"
    >
      <input
        class="govuk-checkboxes__input"
        id="test2"
        type="checkbox"
      />
      <label
        class="govuk-label govuk-checkboxes__label"
        for="test2"
      >
        Label2
      </label>
    </div>
  </div>
</div>
`;
